### PR TITLE
Endpoints

### PR DESCRIPTION
**Pull request by HeDo88TH**
Endpoints are now almost complete: MVP reached.

### CODE REVIEW

**Comment by HeDo88TH** [comment]
I'm fixing the final issues with merge and I'm updating the docs to better reflect the last changes

**Comment by HeDo88TH** [comment]
Checks are failing because we should install ddb in CI pipeline before running tests

File: Registry.Web/Services/Adapters/ObjectsManager.cs [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+                _logger.LogInformation("Bucket created");
+
+            }
+
+            var objInfo = await _objectSystem.GetObjectInfoAsync(bucketName, path);
+
+            if (objInfo == null)
+                throw new NotFoundException($"Cannot find '{path}' in storage provider");
+            
+            await using var memory = new MemoryStream();
```
Comment (pierotofy) [inline Registry.Web/Services/Adapters/ObjectsManager.cs:111]: What if the file is really big?

File: Registry.Ports/DroneDB/Models/DdbObject.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
+﻿using System;
+using GeoJSON.Net;
+using GeoJSON.Net.Feature;
+using GeoJSON.Net.Geometry;
+using Newtonsoft.Json.Linq;
+
+namespace Registry.Ports.DroneDB.Models
+{
+    public class DdbObject
```
Comment (pierotofy) [inline Registry.Ports/DroneDB/Models/DdbObject.cs:9]: This should be renamed `DdbEntry`.

**Comment by pierotofy** [comment]
Making the following request:

`POST http://localhost:5000/ddb/share/init` without auth info returns an empty response (error code is correct, 401)

It should return:

```
{
    "error": "Unauthorized"
}
```

**Comment by pierotofy** [comment]
`POST /ddb/share/init` should be changed a bit:
 - No JSON fields. Should use form data (just like `/users/authenticate`).
 - The only fields it should expose are: [`tag` (optional), `password` (optional)]
 - It's the task of the Registry to parse `tag` (which is a representation of `organization/dataset`). 
 - If no `tag` is specified, the dataset should be placed in the system/public organization and a random dataset UUID should be picked.
 - A password field needs to be added to the dataset model.
 - A user should not be able to specify `slug` manually. These are always computed based on the `name`.

`ddb share [FILES|DIRECTORIES] [FLAGS] [--tag orgname/dataset] [--password <pwd>]`

**Comment by pierotofy** [comment]
This will probably get fixed with the changes proposed above, but for the sake of testing, calling `localhost:5000/ddb/share/init` with:

```
{
    "organization": {
        "id": "test",
        "name": "Test organization",
        "description": "This is a test organization",
        "isPublic": true
    },
    "dataset": {
        "slug": "first",
        "name": "First dataset",
        "description": "This is my first dataset",
        "isPublic": true
    }
}
```

Returns:

```
{
    "error": "Value cannot be null. (Parameter 'source')"
}
```

**Comment by pierotofy** [comment]
We need to be **really** careful about using `POST /ddb/share/upload/{token}/{path/to/file}`. {path/to/file} can contain unicode characters, which would force the client to URL-encode the path.

Perhaps `path` should be just a form data field?

https://stackoverflow.com/questions/2742852/unicode-characters-in-urls

**Comment by pierotofy** [comment]
`POST localhost:5000/ddb/` should accept form data, not JSON.

**Comment by pierotofy** [comment]
`DELETE http://localhost:5000/ddb/test2` should return something like:

```
{
    "success": true
}
```

Or some other confirmation.

**Comment by pierotofy** [comment]
`POST localhost:5000/ddb/test2/ds/aukerman` or even `POST localhost:5000/ddb/test2/ds` return `405 not allowed` or `404 not found`. Note I'm not sure this endpoint will be necessary in the end. Meta endpoints like `ddb/share` might be the only ones that allow the creation of a dataset.

**Comment by pierotofy** [comment]
`PUT http://localhost:5000/ddb/test2` return no content (should return something affermative like `{ "success": true }' or a copy of the newly modified model?

**Comment by pierotofy** [comment]
If I call `POST http://localhost:5000/ddb/` with:

```
{
    "id": 3,
    "name": "Test organization",
    "creationDate": 0
}
```

It explodes. I think a person should be able to create an organization by simply specifying the ID and name (optionally), all other fields should be prepopulated.

**Comment by pierotofy** [comment]
I think these are good starters! Let me know if there are questions. :+1:

File: Registry.Ports/DroneDB/Models/DdbInfo.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        public Feature PointGeometry { get; set; }
+        [JsonProperty("polygon_geom")]
+        public Feature PolygonGeometry { get; set; }
+        [JsonProperty("size")]
+        public int Size { get; set; }
+        [JsonProperty("type")]
+        public DdbInfoType Type { get; set; }
+    }
+
+    public enum DdbInfoType
```
[the remark this answers is not in the document]
Comment (HeDo88TH) [reply Registry.Ports/DroneDB/Models/DdbInfo.cs]: Fixed in ab2495b

File: Registry.Ports/DroneDB/Models/DdbObject.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (HeDo88TH) [reply Registry.Ports/DroneDB/Models/DdbObject.cs:10]: Yes, one for concern.

File: Registry.Web.Test/DdbFactoryTest.cs [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+    [TestFixture]
+    public class DdbFactoryTest
+    {
+        private Mock<IOptions<AppSettings>> _appSettingsMock;
+        private Logger<DdbFactory> _ddbFactoryLogger;
+
+        private const string TestDataFolder = @"Data/Ddb";
+        private const string DdbTestDataFolder = @"Data/DdbTest";
+
+        private const string Test1ArchiveUrl = "https://digipa.it/wp-content/uploads/2020/08/Test1.zip";
```
[the remark this answers is not in the document]
Comment (HeDo88TH) [reply Registry.Web.Test/DdbFactoryTest.cs]: Changed in 432d871

File: Registry.Web.Test/ObjectManagerTest.cs [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+
+        private const string DataFolder = "Data";
+        private const string TestStorageFolder = @"Data/Storage";
+        private const string DdbTestDataFolder = @"Data/DdbTest";
+        private const string StorageFolder = "Storage";
+        private const string DdbFolder = "Ddb";
+
+        private const string BaseTestFolder = "ObjectManagerTest";
+
+        private const string Test1ArchiveUrl = "https://digipa.it/wp-content/uploads/2020/08/Test1.zip";
```
[the remark this answers is not in the document]
Comment (HeDo88TH) [reply Registry.Web.Test/ObjectManagerTest.cs]: Fixed in 432d871

**Comment by HeDo88TH** [comment]
Switched from JSON to form-data as input parameters format for all web services (24ca60a)

File: Registry.Web/Controllers/ObjectsController.cs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        public async Task<IActionResult> Post([FromRoute] string orgId, [FromRoute] string dsId, string path, IFormFile file)
+        {
+            try
+            {
+                _logger.LogDebug($"Objects controller Post('{orgId}', '{dsId}', '{path}', '{file?.FileName}')");
+
+                if (file == null)
+                    return BadRequest(new ErrorResponse("No file uploaded"));
+
+                await using var memory = new MemoryStream();
```
[the remark this answers is not in the document]
Comment (HeDo88TH) [reply Registry.Web/Controllers/ObjectsController.cs:84]: Both #33 and #31 are work in progress

File: Registry.Web/Controllers/ShareController.cs [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        {
+            try
+            {
+
+                _logger.LogDebug($"Share controller Upload('{token}', '{path}', '{file?.FileName}')");
+
+                if (file == null)
+                    return BadRequest(new ErrorResponse("No file uploaded"));
+                
+                await using var memory = new MemoryStream();
```
[the remark this answers is not in the document]
Comment (HeDo88TH) [reply Registry.Web/Controllers/ShareController.cs:61]: Both #33 and #31 are work in progress

File: Registry.Web/Services/Adapters/ObjectsManager.cs (hunk 1 above)
Comment (HeDo88TH) [reply Registry.Web/Services/Adapters/ObjectsManager.cs:111]: Both #33 and #31 are work in progress

File: Registry.Ports/DroneDB/Models/DdbObject.cs (hunk 2 above)
Comment (HeDo88TH) [reply Registry.Ports/DroneDB/Models/DdbObject.cs]: Fixed in cf6a65a

File: Registry.Web/Controllers/ObjectsController.cs [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+using Registry.Web.Models;
+using Registry.Web.Models.DTO;
+using Registry.Web.Services.Ports;
+
+namespace Registry.Web.Controllers
+{
+
+    [Authorize]
+    [ApiController]
+    [Route("ddb/{orgId:alpha}/ds/{dsId:alpha}")]
```
[the remark this answers is not in the document]
Comment (HeDo88TH) [reply Registry.Web/Controllers/ObjectsController.cs]: Fixed in a3140a1

**Comment by HeDo88TH** [comment]
> `DELETE http://localhost:5000/ddb/test2` should return something like:
> 
> ```
> {
>     "success": true
> }
> ```
> 
> Or some other confirmation, instead of an empty response? I know we can look at the status codes, but it might be better to send some content along with the status code.

As stated in https://stackoverflow.com/questions/2342579/http-status-code-for-update-and-delete the correct status code is 204 No Content, that has... no content.

**Comment by pierotofy** [comment]
Mm, maybe we can return `200` instead then?

**Comment by HeDo88TH** [comment]
Let's stick to the standard. If a DELETE fails It will return 4XX with error message, otherwise everything went well.

File: Registry.Adapters.DroneDB/Ddb.cs [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+            Directory.CreateDirectory(tempPath);
+
+            var destFolder = Path.GetDirectoryName(path);
+            Directory.CreateDirectory(destFolder);
+
+            var res = RunCommand($"{InitCommand} -d \"{tempPath}\"");
+
+            // This way we preserve the ACL
+            File.WriteAllBytes(path, File.ReadAllBytes(Path.Combine(tempPath, ".ddb", "dbase.sqlite")));
```
[the remark this answers is not in the document]
Comment (HeDo88TH) [reply Registry.Adapters.DroneDB/Ddb.cs]: Fixed in bbe15fe

**Comment by HeDo88TH** [comment]
> This will probably get fixed with the changes proposed above, but for the sake of testing, calling `localhost:5000/ddb/share/init` with:
> 
> ```
> {
>     "organization": {
>         "id": "test",
>         "name": "Test organization",
>         "description": "This is a test organization",
>         "isPublic": true
>     },
>     "dataset": {
>         "slug": "first",
>         "name": "First dataset",
>         "description": "This is my first dataset",
>         "isPublic": true
>     }
> }
> ```
> 
> Returns:
> 
> ```
> {
>     "error": "Value cannot be null. (Parameter 'source')"
> }
> ```

Modified in 08d102d

**Comment by HeDo88TH** [comment]
All the issues and proposals have been addressed. I'm going to write some end-to-end tests to ensure that the overall behaviour is ok, then we can merge it @pierotofy

**Comment by pierotofy** [comment]
Excellent! I will be testing the new changes after the tests are pushed.

**Comment by pierotofy** [comment]
`POST /ddb/share/init` with:

`tag=test/test` (where test is a valid org):

Returns:

```
{
    "error": "Invalid dataset"
}
```

Shouldn't I be allowed to initiate a share upload to the admin organization?

**Comment by pierotofy** [comment]
`POST localhost:5000/ddb/share/upload/CgF0IsVVIWkXEDEKFiHOUmg9FQiFYVe6/DJI.JPG`

With a single image (JPG).

Returns:

```
{
    "type": "https://tools.ietf.org/html/rfc7231#section-6.5.13",
    "title": "Unsupported Media Type",
    "status": 415,
    "traceId": "|6452c39b-49aed024ec18ad75."
}
```

If I choose `binary` from Postman. I think this is wrong (docs not clear).

This should be a multipart/form-data file upload (what's the key for the file?)

Anyway, if I try form-data upload I get:

```
{
    "error": "Sqlite database not found"
}
```

I also think we should place the `path` in the form-data body instead of the URL.

**Comment by pierotofy** [comment]
Let's not add more things to this PR, but for the very near future: https://github.com/DroneDB/Registry/issues/38

**Comment by HeDo88TH** [comment]
I rolled back b360492 and 25f1ca8 because there is no need to return an object when the HTTP code is already stating it.
I fixed 5aac056, some tests were broken and some logic too.
I added back the fields DatasetName and DatasetDescription to ShareInitDto as optional if the clients wants to specify more details about the dataset.
I added a new TagDto class that handles the parsing of tags like "organization/dataset" and I reorganized some utils methods in extenders methods.

**Comment by pierotofy** [comment]
Maybe it's just my environment, but I still cannot get a file upload going:

`POST /ddb/share/init`

Passing: `tag=test2/test2` as form-data. The org `test2` exists.

```
      Dataset 'test2' not found, creating it
fail: Registry.Web.Controllers.ShareController[0]
      Exception in Share controller Init('Registry.Web.Models.DTO.ShareInitDto')
Registry.Web.Exceptions.NotFoundException: Cannot find dataset
   at Registry.Web.Services.Adapters.WebUtils.GetDatasetAndCheck(String orgSlug, String dsSlug, Boolean retNullIfNotFound) in D:\RegistryApp\Registry.Web\Services\Adapters\WebUtils.cs:line 81
   at Registry.Web.Services.Adapters.ShareManager.Initialize(ShareInitDto parameters) in D:\RegistryApp\Registry.Web\Services\Adapters\ShareManager.cs:line 123
   at Registry.Web.Controllers.ShareController.Init(ShareInitDto parameters) in D:\RegistryApp\Registry.Web\Controllers\ShareController.cs:line 35
```

**Comment by pierotofy** [comment]
This is still not working. :(

`POST /ddb/share/upload/<validToken>/IMG_20160904_130804.jpg`

```
{
"error": "Invalid path"
}
```

Side note, path is still encoded in the URL. I thought we addressed this in a previous comment during this review.

>> I also think we should place the path in the form-data body instead of the URL.

**Review by pierotofy — CHANGES_REQUESTED** [review]
1) Make sure nominal test paths work
2) Make sure `path` is encoded in the form-data body, not URL

**Comment by HeDo88TH** [comment]
> 1. Make sure nominal test paths work

I tested with postman and It works on my machine. I added logging to a text file to improve issue tracking (d013f57), when something breaks attach the log file to better identify the issue.

> 2. Make sure `path` is encoded in the form-data body, not URL

Addressed in d2d8fa7, I will update the docs as soon as possible.

**Comment by pierotofy** [comment]
:+1: 

Excellent!

I've made a mistake while testing; (was using the wrong IP address in postman).

**Pull request merged**